library/manageiq.py

```diff
@@ -118,11 +118,21 @@ def update_required(self, provider_id, endpoints):
             self.module.fail_json(msg="Failed to get provider data. Error: %s" % e)
 
         def host_port(endpoint):
-            return (endpoint['hostname'], int(endpoint['port']))
+            return {'hostname': endpoint['hostname'], 'port': int(endpoint['port'])}
 
         desired_by_role = {e['endpoint']['role']: host_port(e['endpoint']) for e in endpoints}
         result_by_role = {e['role']: host_port(e) for e in result['endpoints']}
-        return desired_by_role != result_by_role
+        if result_by_role == desired_by_role:
+            return {}
+        updated = {role: {k: v for k, v in ep.items()
+                          if k not in result_by_role[role] or v != result_by_role[role][k]}
+                   for role, ep in desired_by_role.items()
+                   if role in result_by_role and ep != result_by_role[role]}
+        added = {role: ep for role, ep in desired_by_role.items()
+                 if role not in result_by_role}
+        removed = {role: ep for role, ep in result_by_role.items()
+                   if role not in desired_by_role}
+        return {"Updated": updated, "Added": added, "Removed": removed}
```

Comment:
I'm thinking something like:

    result_keys = set(result_by_role)
    desired_keys = set(desired_by_role)
    added = sorted(desired_keys - result_keys)
    removed = sorted(result_keys - desired_keys)
    updated = sorted(role for role in (desired_keys & result_keys)
                     if desired_by_role[role] != result_by_role[role])

This omits the host port details, but I'm not sure how useful those are.